app/src/main/kotlin/com/simplemobiletools/musicplayer/extensions/Activity.kt

```diff
@@ -70,40 +69,61 @@ fun Activity.playNextInQueue(track: Track, callback: () -> Unit) {
 }
 
 fun BaseSimpleActivity.deleteTracks(tracks: List<Track>, callback: () -> Unit) {
-    val uri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
-    if (isRPlus()) {
-        val (tracksWithoutId, tracksWithId) = tracks.partition { (it.flags and FLAG_MANUAL_CACHE) != 0 }
-
-        val uris = tracksWithId.map { ContentUris.withAppendedId(uri, it.mediaStoreId) } as ArrayList<Uri>
-        val contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
-        uris += tracksWithoutId.mapNotNull { getMediaContent(it.path, contentUri) }
+    try {
+        audioHelper.deleteTracks(tracks)
+        audioHelper.removeInvalidAlbumsArtists()
+    } catch (ignored: Exception) {
+    }
 
-        deleteSDK30Uris(uris) { success ->
-            if (success) {
-                removeQueueItems(tracks)
+    val contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
+    maybeRescanTrackPaths(tracks) { tracksToDelete ->
+        if (tracksToDelete.isNotEmpty()) {
+            if (isRPlus()) {
+                val uris = tracksToDelete.map { ContentUris.withAppendedId(contentUri, it.mediaStoreId) }
+                deleteSDK30Uris(uris) { success ->
+                    if (success) {
+                        removeQueueItems(tracksToDelete)
+                        EventBus.getDefault().post(Events.RefreshFragments())
+                        callback()
+                    } else {
+                        toast(R.string.unknown_error_occurred)
+                    }
+                }
+            } else {
+                tracksToDelete.forEach { track ->
+                    try {
+                        val where = "${MediaStore.Audio.Media._ID} = ?"
+                        val args = arrayOf(track.mediaStoreId.toString())
+                        contentResolver.delete(contentUri, where, args)
+                        File(track.path).delete()
+                    } catch (ignored: Exception) {
+                    }
+                }
+
+                removeQueueItems(tracksToDelete)
                 EventBus.getDefault().post(Events.RefreshFragments())
                 callback()
-            } else {
-                toast(R.string.unknown_error_occurred)
             }
         }
-        return
     }
+}
 
-    tracks.forEach { track ->
-        try {
-            val where = "${MediaStore.Audio.Media._ID} = ?"
-            val args = arrayOf(track.mediaStoreId.toString())
-            contentResolver.delete(uri, where, args)
-            audioHelper.deleteTrack(track.mediaStoreId)
-            File(track.path).delete()
-        } catch (ignored: Exception) {
+private fun Activity.maybeRescanTrackPaths(tracks: List<Track>, callback: (tracks: List<Track>) -> Unit) {
+    val tracksWithoutId = tracks.filter { it.mediaStoreId == 0L || it.flags and FLAG_MANUAL_CACHE != 0 }
```

Comment:
add some brackets here to make it clearer

Reply:
added